lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueAutoPingLockStrategy.java

```diff
@@ -0,0 +1,277 @@
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.esbtools.eventhandler.lightblue;
+
+import com.redhat.lightblue.client.LightblueException;
+import com.redhat.lightblue.client.Locking;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Uses lightblue's locking APIs with TTL, automatically pinging the lock in the background until it
+ * is released.
+ *
+ * <p>Lightblue locks without TTLs are unsafe. If a client crashes, the lock will not be released.
+ * However, locks with TTLs may expire prematurely. By pinging the lock in a separate thread
+ * periodically, we can substantially lesson the likelihood of unintentionally losing the lock.
+ *
+ * <p>Consumers are expected to check the lock at critical points to ensure it did not expire by
+ * calling {@link LockedResource#ensureAcquiredOrThrow(String)}.
+ */
+public class LightblueAutoPingLockStrategy implements LockStrategy {
+    private final Locking locking;
+    private final Duration tryAcquireInterval;
+    private final Duration autoPingInterval;
+    private final Duration timeToLive;
+
+    /**
+     * Same as {@link #LightblueAutoPingLockStrategy(Locking, Duration, Duration, Duration)}
+     * except the {@code timeToLive} duration defaults to 5 times the {@code autoPingInterval}.
+     */
+    public LightblueAutoPingLockStrategy(Locking locking, Duration tryAcquireInterval,
+            Duration autoPingInterval) {
+        this(locking, tryAcquireInterval, autoPingInterval, autoPingInterval.multipliedBy(5));
+    }
+
+    /**
+     * Validates {@code timeToLive} is greater than the {@code autoPingInterval}.
+     * @param locking
+     * @param tryAcquireInterval If the asked resources cannot be acquire, will wait this duration
+     *                           before trying again.
+     * @param autoPingInterval Amount of time in between automatic pings of acquired locks.
+     * @param timeToLive Time until locks automatically expire. Should be [much] larger than the
+     *                   {@code autoPingInterval} to ensure locks do not accidentally expire.
+     */
+    public LightblueAutoPingLockStrategy(Locking locking, Duration tryAcquireInterval,
+            Duration autoPingInterval, Duration timeToLive) {
+        this.locking = locking;
+        this.tryAcquireInterval = tryAcquireInterval;
+        this.autoPingInterval = autoPingInterval;
+        this.timeToLive = timeToLive;
+
+        if (timeToLive.compareTo(autoPingInterval) <= 0) {
```

Comment:
Counter arguments:

- Large TTL delays recovery (relates to your point about recovery per entity). There's not really a good way to determine a good TTL without background ping. Auto pinging in separate thread keeps TTL relatively low.
- It doesn't prevent losing when we want to lose it. We still release the lock and also stop the thread when doing so.
- If the node dies, so does the pinging thread, allowing lock to expire.

Am I missing something? I thought we talked about this kind of impl before.

In the future, we would like to not use lightblue for synchronizing distributed apps and instead look at something like ZooKeeper. I'd do this now, but this will add time.